### PR TITLE
[15012]: Bug fix, small refactor and tests in TlsConfiguration

### PR DESCRIPTION
**Pull request by antonrv**
* Bug fix according to reported issue: https://github.com/eProsima/DDS-Router/issues/243

* Small refactor to simplify TlsConfiguration architecture (bloated class hierarchy, unneeded polymorphism)

* New tests for previously untested scenarios

### CODE REVIEW

File: ddsrouter_core/src/cpp/types/security/tls/TlsConfiguration.cpp [hunk 1]
```diff
@@ -18,47 +18,167 @@
...
 {
-    return false;
+    return this->kind_ != TlsKind::inactive;
+}
+
+template <TlsKind Kind>
+bool TlsConfiguration::compatible() const noexcept
+{
+    return this->kind_ != TlsKind::inactive &&
+           (Kind == this->kind_ or (this->kind_ == TlsKind::both && Kind != TlsKind::inactive));
```
Comment (jparisu) [inline ddsrouter_core/src/cpp/types/security/tls/TlsConfiguration.cpp:113]: Be consistent using `or` or `||`.
*Note: I think `or` is not allowed in windows*

**Review by jparisu — CHANGES_REQUESTED** [review]
Cool and elegant Implementation :+1: 
Just minor suggestions.

File: ddsrouter_core/include/ddsrouter_core/types/security/tls/TlsConfiguration.hpp [hunk 2]
```diff
@@ -29,41 +29,147 @@ namespace core {
...
+    DDSROUTER_CORE_DllAPI bool is_active() const noexcept;
+
+    /**
+     * @brief Check configuration kind compatibility
+     *
+     * @tparam Kind TlsConfiguration kind
+     *
+     * @return Whether configuration is compatible with parameter \c TlsKind
+     */
+    template <TlsKind Kind>
```
[the remark this answers is not in the document]
Comment (antonrv) [reply ddsrouter_core/include/ddsrouter_core/types/security/tls/TlsConfiguration.hpp:113]: Less chances to call it badly. Also implementation `::compatible<TlsKind::inactive>` is deliberately not implemented as it is a forbidden value, so incorrect usage will be noted at compile-time

File: ddsrouter_core/src/cpp/types/security/tls/TlsConfiguration.cpp (hunk 1 above)
Comment (antonrv) [reply ddsrouter_core/src/cpp/types/security/tls/TlsConfiguration.cpp:113]: right, win complains about it and I forgot to remove all of them

**Review by jparisu — COMMENTED** [review]
Everything LGTM,
Windows fails linking new methods nevertheless.

**Pull request merged**